inject-deps.escript

```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/env escript
+
+%% This script injects implicit relup dependencies for emqx applications.
+%%
+%% By 'implicit', it means that it is not feasible to define application
+%% dependencies in .app.src files.
+%%
+%% For instance, during upgrade/downgrade, emqx_dashboard usually requires
+%% a restart after (but not before) all plugins are upgraded (and maybe
+%% restarted), however, the dependencies are not resolvable at build time
+%% when relup is generated.
+%%
+%% This script is to be executed after compile, with the profile given as the
+%% first argument. For each dependency overlay, it modifies the .app file to
+%% have the 'relup_deps' list extended.
+
+-mode(compile).
+
+usage() ->
+  "Usage: " ++ escript:script_name() ++ " emqx|edge".
+
+-type app() :: atom().
+-type deps_overlay() :: {re, string()} | app().
+
+%% deps/0 returns the dependency overlays.
+%% {re, Pattern} to match application names using regexp pattern
+-spec deps(string()) -> [{app(), [deps_overlay()]}].
+deps(_Profile) ->
+  [ {emqx_dashboard, [{re, "emqx_.*"}]}
```

Comment:
Does the application need to support wildcards as well? And I think a full dependency relationship should look like this.

```
[{emqx_dashboard, {re, "emqx_"}}],
 {emqx_management, {re, "emqx_"}},
 {{re, "emqx_*"}, emqx},
 {{re, "emqx_auth_*"}, emqx_passwd}
]
```

Reply:
great idea.
to workaround circular dependency, we can add an 'exclude' pattern.

Reply:
Cool!